File: 68.py
class Solution(object):
    def fullJustify(self, words, maxWidth):
        index = 0
        res = []
        while index < len(words):
            lastline = (index == len(words) - 1)
            arr = [words[index]]
            tot = len(words[index])
            index += 1
            while index < len(words) and tot + len(arr) + len(words[index]) <= maxWidth:
                tot += len(words[index])
                arr.append(words[index])
                if index == len(words)-1: lastline = True
                index += 1
            s = ""
            if len(arr) == 1: 
                s = arr[0] + (" " * (maxWidth-tot))
            elif lastline:
                s = ' '.join(arr) + (" " * (maxWidth-tot-(len(arr)-1)))
            else:
                quotient, remainder = divmod(maxWidth-tot, len(arr)-1)
                for i in range(len(arr)-1):
                    s += arr[i] + (" " * quotient)
                    if i < remainder: s += " "
                s += arr[-1]
            res.append(s)
        return res

File: 68.js
/**
 * @param {string[]} words
 * @param {number} maxWidth
 * @return {string[]}
 */
String.prototype.leftJustify = function( length, char ) {
    var fill = [];
    while ( fill.length + this.length < length ) {
      fill[fill.length] = char;
    }
    return this + fill.join('');
}

var fullJustify = function(words, maxWidth) {
    let index = 0
    let res = []
    while (index < words.length) {
        let lastline = (index == words.length - 1)
        let arr = [words[index]]
        let tot = words[index].length
        index++
        for (; index < words.length && tot + arr.length + words[index].length <= maxWidth; index++) {
            tot += words[index].length
            arr.push(words[index])
            if (index == words.length - 1) lastline = true
        }
        if (arr.length == 1 || lastline) {
            res.push(arr.join(' ').leftJustify(maxWidth, ' '))
        }
        else {
            for (let i = 0; i < maxWidth - tot; i++) arr[i % (arr.length-1 || 1)] += ' '
            res.push(arr.join(''))
        }
    }
    return res
}